File: python/source/cate_set/province_count_547.py
class Node:
    def __init__(self, val:int, parent):
        self.val = val
        self.parent = parent
        self.factor = 1
    def __str__(self):
        return "self: {}, parent: {}".format(self.val, self.parent.val if self.parent else None)

class UnionFindSet(object):
    def __init__(self, size):
        self.nodes = [Node(i, None) for i in range(size)]
        for node in self.nodes:
            node.parent = node
        self.size = size

    def parent(self, node:Node):
        # 找到根节点
        node_ptr = node
        while node_ptr.parent != node_ptr:
            node_ptr = node_ptr.parent
        root = node_ptr
        # 路径压缩
        node_ptr = node
        while node_ptr != root:
            parent = node_ptr.parent
            node_ptr.parent = root
            node_ptr = parent
        return root

    def merge(self, node1, node2):
        p1 = self.parent(node1)
        p2 = self.parent(node2)
        p2.parent = p1

    def build(self, isConnected):
        if not isConnected:
            return
        for row in range(self.size):
            for col in range(row+1, self.size):
                if isConnected[row][col] == 0:
                    continue
                self.merge(self.nodes[row], self.nodes[col])

    def count(self):
        return len(set([self.parent(node) for node in self.nodes]))




class Solution:
    def findCircleNum(self, isConnected) -> int:
        if not isConnected:
            return 0
        unionSet = UnionFindSet(len(isConnected))
        unionSet.build(isConnected)
        return unionSet.count()


if __name__ == '__main__':
    so = Solution()
    isConnected = [[1,0,0,1],[0,1,1,0],[0,1,1,1],[1,0,1,1]]
    res = so.findCircleNum(isConnected)
    print(res)

File: java/src/main/java/UnionSet/ProvinceCount_547.java
package UnionSet;

/**
 * 有 n 个城市，其中一些彼此相连，另一些没有相连。如果城市 a 与城市 b 直接相连，且城市 b 与城市 c 直接相连，
 * 那么城市 a 与城市 c 间接相连。
 *
 * 省份 是一组直接或间接相连的城市，组内不含其他没有相连的城市。
 *
 * 给你一个 n x n 的矩阵 isConnected ，其中 isConnected[i][j] = 1 表示第 i 个城市和第 j 个城市直接相连，
 * 而 isConnected[i][j] = 0 表示二者不直接相连。
 *
 * 返回矩阵中 省份 的数量。
 *
 */
public class ProvinceCount_547 {
    public int findCircleNum(int[][] isConnected) {
        int count = isConnected.length;
        UnionFind uf = new UnionFind(count);
        for(int i = 0; i < count; i++){
            for(int j = i+1; j < count; j++){
                if(isConnected[i][j] == 1){
                    uf.union(i, j);
                }
            }
        }
        return uf.setCount();
    }
}
